cpp/src/arrow/compute/kernels/aggregate-benchmark.cc

```diff
@@ -0,0 +1,349 @@
+static CpuInfo* cpu_info = CpuInfo::GetInstance();
+
+static const int64_t kL1Size = cpu_info->CacheSize(CpuInfo::L1_CACHE);
+static const int64_t kL2Size = cpu_info->CacheSize(CpuInfo::L2_CACHE);
+static const int64_t kL3Size = cpu_info->CacheSize(CpuInfo::L3_CACHE);
+
+namespace BitUtil = arrow::BitUtil;
+using arrow::internal::BitmapReader;
+
+template <typename T>
+struct SumState {
+  using ValueType = T;
+
+  SumState() : total(0), valid_count(0) {}
+
+  T total = 0;
+  int64_t valid_count = 0;
+};
+
+template <typename T>
+struct Traits {};
+
+template <>
+struct Traits<int64_t> {
+  using ArrayType = typename CTypeTraits<int64_t>::ArrayType;
+  static constexpr int64_t null_sentinel = std::numeric_limits<int64_t>::lowest();
+
+  static void FixSentinel(std::shared_ptr<ArrayType>& array) {
+    auto data = array->data();
+    for (int64_t i = 0; i < array->length(); i++)
+      if (array->IsNull(i)) {
+        int64_t* val_ptr = data->GetMutableValues<int64_t>(1, i);
+        *val_ptr = null_sentinel;
+      }
+  }
+
+  static inline bool IsNull(int64_t val) { return val == null_sentinel; }
+
+  static inline bool NotNull(int64_t val) { return val != null_sentinel; }
+};
+
+template <typename T>
+struct Summer {
+ public:
+  using ValueType = T;
+  using ArrowType = typename CTypeTraits<T>::ArrowType;
+};
+
+template <typename T>
+struct SumNoNulls : public Summer<T> {
+  using ArrayType = typename CTypeTraits<T>::ArrayType;
+
+  static void Sum(const ArrayType& array, SumState<T>* state) {
+    SumState<T> local;
+
+    const auto values = array.raw_values();
+    for (int64_t i = 0; i < array.length(); ++i) {
+      local.total += values[i];
+    }
+
+    local.valid_count = array.length();
+    *state = local;
+  }
+};
+
+template <typename T>
+struct SumNoNullsUnrolled : public Summer<T> {
+  using ArrayType = typename CTypeTraits<T>::ArrayType;
+
+  static void Sum(const ArrayType& array, SumState<T>* state) {
+    SumState<T> local;
+
+    const auto values = array.raw_values();
+    const auto length = array.length();
+    const int64_t length_rounded = BitUtil::RoundDown(length, 8);
+    for (int64_t i = 0; i < length_rounded; i += 8) {
+      local.total += values[i + 0] + values[i + 1] + values[i + 2] + values[i + 3] +
+                     values[i + 4] + values[i + 5] + values[i + 6] + values[i + 7];
+    }
+
+    for (int64_t i = length_rounded; i < length; ++i) {
+      local.total += values[i];
+    }
+
+    local.valid_count = length;
+
+    *state = local;
+  }
+};
+
+template <typename T>
+struct SumSentinel : public Summer<T> {
+  using ArrayType = typename CTypeTraits<T>::ArrayType;
+
+  static void Sum(const ArrayType& array, SumState<T>* state) {
+    SumState<T> local;
+
+    const auto values = array.raw_values();
+    const auto length = array.length();
+    for (int64_t i = 0; i < length; i++) {
+      // NaN is not equal to itself
+      local.total += values[i] * Traits<T>::NotNull(values[i]);
```

Comment:
The previous version was faster, clang-6 and onward auto-vectorize this perfectly (even with just O3): https://godbolt.org/z/v5nFNM . The result would be even better with AVX512 enabled (see https://godbolt.org/z/LFcQJv and  https://llvm.org/devmtg/2015-04/slides/MaskedIntrinsics.pdf).